module/os/windows/spl/spl-kmem.c

```diff
@@ -0,0 +1,6628 @@
+			TraceEvent(TRACE_ERROR, "redzone violation: write past"
+			    " end of buffer\n");
+			break;
+
+		case KMERR_BADADDR:
+			TraceEvent(TRACE_ERROR, "invalid free: buffer not in"
+			    " cache\n");
+			break;
+
+		case KMERR_DUPFREE:
+			TraceEvent(TRACE_ERROR, "duplicate free: buffer freed"
+			    " twice\n");
+			break;
+
+		case KMERR_BADBUFTAG:
+			TraceEvent(TRACE_ERROR, "boundary tag corrupted\n");
+			TraceEvent(TRACE_ERROR, "SPL: bcp ^ bxstat = %lx, "
+			    "should be %lx\n",
+			    (intptr_t)btp->bt_bufctl ^ btp->bt_bxstat,
+			    KMEM_BUFTAG_FREE);
+			break;
+
+		case KMERR_BADBUFCTL:
+			TraceEvent(TRACE_ERROR, "bufctl corrupted\n");
+			break;
+
+		case KMERR_BADCACHE:
+			TraceEvent(TRACE_ERROR, "buffer freed to wrong "
+			    "cache\n");
+			TraceEvent(TRACE_ERROR, "SPL: buffer was allocated"
+			    " from %s,\n", cp->cache_name);
+			TraceEvent(TRACE_ERROR, "SPL: caller attempting free"
+			    " to %s.\n", cparg->cache_name);
+			break;
+
+		case KMERR_BADSIZE:
+			TraceEvent(TRACE_ERROR, "bad free: free size (%u) !="
+			    " alloc size (%u)\n",
+			    KMEM_SIZE_DECODE(((uint32_t *)btp)[0]),
+			    KMEM_SIZE_DECODE(((uint32_t *)btp)[1]));
+			break;
+
+		case KMERR_BADBASE:
+			TraceEvent(TRACE_ERROR, "bad free: free address"
+			    " (%p) != alloc address (%p)\n", bufarg, buf);
+			break;
+	}
+
+	dprintf("SPL: buffer=%p  bufctl=%p  cache: %s\n",
+	    bufarg, (void *)bcp, cparg->cache_name);
+
+	if (bcp != NULL && (cp->cache_flags & KMF_AUDIT) &&
+	    error != KMERR_BADBUFCTL) {
+		int d;
+		timestruc_t ts = {0, 0};
+		kmem_bufctl_audit_t *bcap = (kmem_bufctl_audit_t *)bcp;
+
+		hrt2ts(kmem_panic_info.kmp_timestamp - bcap->bc_timestamp, &ts);
+		dprintf("SPL: previous transaction on buffer %p:\n", buf);
+		dprintf("SPL: thread=%p  time=T-%ld.%09ld  slab=%p  cache: "
+		    "%s\n", (void *)bcap->bc_thread, ts.tv_sec, ts.tv_nsec,
+		    (void *)sp, cp->cache_name);
+		for (d = 0; d < MIN(bcap->bc_depth, KMEM_STACK_DEPTH); d++) {
+			dprintf("   : %p\n, ", (void *)bcap->bc_stack[d]);
+		}
+	}
+
+	if (kmem_panic > 0) {
+		extern  void IODelay(unsigned microseconds); // <IOKit/IOLib.h?
+		DbgBreakPoint();
+		panic("kernel heap corruption detected");
+	}
+
+	kmem_logging = 1;	/* resume logging */
+}
+
+static kmem_log_header_t *
+kmem_log_init(size_t logsize)
+{
+	kmem_log_header_t *lhp;
+	int nchunks = 4 * max_ncpus;
+	size_t lhsize = (size_t)&((kmem_log_header_t *)0)->lh_cpu[max_ncpus];
+	int i;
+
+	/*
+	 * Make sure that lhp->lh_cpu[] is nicely aligned
+	 * to prevent false sharing of cache lines.
+	 */
+	lhsize = P2ROUNDUP(lhsize, KMEM_ALIGN);
+	lhp = vmem_xalloc(kmem_log_arena, lhsize, 64, P2NPHASE(lhsize, 64), 0,
+	    NULL, NULL, VM_SLEEP);
+	memset(lhp, 0, lhsize);
+
+	mutex_init(&lhp->lh_lock, NULL, MUTEX_DEFAULT, NULL);
+	lhp->lh_nchunks = nchunks;
+	lhp->lh_chunksize = P2ROUNDUP(logsize / nchunks + 1, PAGESIZE);
+	lhp->lh_base = vmem_alloc_impl(kmem_log_arena,
+	    lhp->lh_chunksize * nchunks, VM_SLEEP);
```

Comment:
Although we should not be doing allocations that are >=4GB in size, if we ever do, this will underallocate. It should be given a typecast to `(size_t)` as a defensive measure to harden against future changes that will try to allocate big things.